File: binarySearch_problems/ArrangeCoins.cpp

// Problem link - https://leetcode.com/problems/arranging-coins/description/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int arrangeCoins(int n) {
        long s = 1, e = n;
        long ans = 1;

        while(s <= e){
            long mid = s + (e - s)/2;
            long val = (mid * (mid+1))/2;
            if(val <= n){
                // get first potential ans, 
                ans = mid;
                // as we have find maximum number of rows (i.e mid), hence we discard left half of array
                s = mid + 1;
            }else{
                e = mid - 1;
            }
        }
        return (int) ans;
    }
};

File: BinarySearch_problems/ArrangeCoins.java

// Problem link - https://leetcode.com/problems/arranging-coins/description/

package BinarySearch_problems;

public class ArrangeCoins {
    public int arrangeCoins(long n) {
        long s = 1, e = n;
        long ans = 1;

        while (s <= e) {
            long mid = s + (e - s) / 2;
            long val = (mid * (mid + 1)) / 2;

            if (val <= n) {
                // we get the first potential ans in val
                ans = mid;
                // if we find potential ans then we can better ans in right half of array,
                // because we have of find maximum number of i.e k (mid) such k<=n
                // discard left half of array
                s = mid + 1;
            } else {
                e = mid - 1;
            }
        }
        return (int)ans;
    }
}
